src/pathfinding/grid_pathfinder.jl

```diff
@@ -0,0 +1,362 @@
+        Dict{Int,Path{D}}(),
+        dims,
+        neighborhood,
+        admissibility,
+        walkable,
+        metric,
+    )
+end
+
+moore_neighborhood(D) = [
+    CartesianIndex(a)
+    for a in Iterators.product([-1:1 for φ in 1:D]...) if a != Tuple(zeros(Int, D))
+]
+
+function vonneumann_neighborhood(D)
+    hypercube = CartesianIndices((repeat([-1:1], D)...,))
+    [β for β ∈ hypercube if LinearAlgebra.norm(β.I) == 1]
+end
+
+function Base.show(io::IO, pathfinder::AStar{D,P,M}) where {D,P,M}
+    periodic = P ? "periodic, " : ""
+    moore = M ? "moore, " : ""
+    s = "A* in $(D) dimensions. $(periodic)$(moore)ϵ=$(pathfinder.admissibility), metric=$(pathfinder.cost_metric)"
+    print(io, s)
+end
+
+"""
+    position_delta(pathfinder::AStar{D}, from::NTuple{Int,D}, to::NTuple{Int,D})
+Returns the absolute difference in coordinates between `from` and `to` taking into account
+periodicity of `pathfinder`.
+"""
+position_delta(pathfinder::AStar{D,true}, from::Dims{D}, to::Dims{D}) where {D} =
+    min.(abs.(to .- from), pathfinder.grid_dims .- abs.(to .- from))
+
+position_delta(pathfinder::AStar{D,false}, from::Dims{D}, to::Dims{D}) where {D} =
+    abs.(to .- from)
+
+"""
+    delta_cost(pathfinder::AStar{D}, metric::M, from, to) where {M<:CostMetric}
+Calculate an approximation for the cost of travelling from `from` to `to` (both of
+type `NTuple{N,Int}`. Expects a return value of `Float64`.
+"""
+function delta_cost(
+    pathfinder::AStar{D,periodic,true},
+    metric::DirectDistance{D},
+    from::Dims{D},
+    to::Dims{D},
+) where {D,periodic}
+    delta = collect(position_delta(pathfinder, from, to))
+
+    sort!(delta)
+    carry = 0
+    hdist = 0
+    for i in D:-1:1
+        hdist += metric.direction_costs[i] * (delta[D+1-i] - carry)
+        carry = delta[D+1-i]
+    end
+    return hdist
+end
+
+function delta_cost(
+    pathfinder::AStar{D,periodic,false},
+    metric::DirectDistance{D},
+    from::Dims{D},
+    to::Dims{D},
+) where {D,periodic}
+    delta = position_delta(pathfinder, from, to)
+
+    return sum(delta) * metric.direction_costs[1]
+end
+
+delta_cost(
+    pathfinder::AStar{D},
+    metric::Chebyshev{D},
+    from::Dims{D},
+    to::Dims{D},
+) where {D} = max(position_delta(pathfinder, from, to)...)
+
+delta_cost(
+    pathfinder::AStar{D},
+    metric::HeightMap{D},
+    from::Dims{D},
+    to::Dims{D},
+) where {D} =
+    delta_cost(pathfinder, metric.base_metric, from, to) +
+    abs(metric.hmap[from...] - metric.hmap[to...])
+
+delta_cost(pathfinder::AStar{D}, from::Dims{D}, to::Dims{D}) where {D} =
+    delta_cost(pathfinder, pathfinder.cost_metric, from, to)
+
+struct GridCell
+    f::Int
+    g::Int
+    h::Int
+end
+
+GridCell(g::Int, h::Int, admissibility::AbstractFloat) =
+    GridCell(round(Int, g + (1 + admissibility) * h), g, h)
+
+GridCell() = GridCell(typemax(Int), typemax(Int), typemax(Int))
+
+"""
+    find_path(pathfinder::AStar{D}, from::NTuple{D,Int}, to::NTuple{D,Int})
```

Comment:
The call signature of this function must become `find_path(space_or_model, from, to)`. Remember: a user never accesses explicitly `AStar`.

Reply:
Actually, does `find_path` need to be part of the public API? I don't think so. Please do not export it and do not put it in the API documentation.

Reply:
I think so. I'm interested in the shortest path to the lowest place on a heightmap. There happens to be 3 lowest places. Which one is closest? How do I set my target?

Reply:
What does the `set_path!` function do in this case?

Reply:
I don't think we have a `set_path!`? You mean `set_target!`? If so, this is the *usual* method we expect people to use. They have a location the agent needs to get to, set that, make a path, done.

We leave `find_path` open so that users can check for paths between arbitrary points when they don't know exactly which position is the target position yet.

Reply:
I'm still not sold. Nowhere it is documented what a `Path` is , not how a user should use it. Where in the public API is the output of `find_path` used? 

What can a user do with this output?

Reply:
> I think so. I'm interested in the shortest path to the lowest place on a heightmap. There happens to be 3 places that have the same minimal value. Which one is closest? How do I set my target?

Currently the `find_path` function doesn't tell which is closest. Is is necessary that the closest is the one with the shortest path? I'm not sure of this, but I'll give it a thought.

Reply:
There is a house and it's on fire, there are multiple exits, I leave my room and head for the closest exit. The fire has blocked me! Where do I go now? This is a very standard situation in crowd dynamics. If we can solve this with `set_target!` and some already existing methods without `find_path`, then I'm OK that we drop it from public.

Reply:
It seems to me that your scenario needs to call dynamically `set_target!` after chancing the `walkable` property of the house space.

Reply:
Changing the walk-able property to disallow the closest exit still does not tell my agent which of the other say, two exits is the best one to go to. 
```julia
space = GridSpace((10, 10))
closest_exit = (1, 3)
other_exits = [(10, 5), (4, 10)]
#etc
agent.pos = (2, 5)
set_target!(agent, closest_exit) 
move_along_path!(agent, model) # agent.pos now (1,5)
# fire burns into hallway, walkable changes, 
walkmap(model)[1,4] = false
# Oh no!
# Find best pathway without find_path
min_path = (Inf, 0)
for exit in exits
    set_target!(agent, exit])
    path_length =  length(model.space.pathfinder.agent_paths[agent.id])
    if path_length < min_path[1]
        min_path = (path_length, exit)
    end
end
set_target!(agent, min_path[2])
```
Solution sets the target three times since we don't know beforehand which one is the quickest.
```julia
# with find_path
path_lengths = map(exit -> length(find_path(model.space.pathfinder, agent.pos, exit)), exits)
idx = findmin(path_lengths)[2]
set_target!(agent, exits[idx!)
```

Reply:
Actually, this would still do the path three times, but it doesn't involve trying to access undocumented structures within the pathfinder struct at least.

Reply:
This is interesting. The current API works to find the shortest path from A to B, assuming you know B. However, here the objective seems to be to find B.
It could be an interesting concept to allow for multiple target locations, and I don't see why it can't be done. All the algorithm has to do is change the termination condition, and take the minimum of the costs to all the targets at each step. It might be worth profiling to see if these changes are actually better than just creating a method `set_targets!` that runs `find_path` as many times as required.

Reply:
Maybe `set_best_target!` with some stopping condition function defaulting to `minimum` perhaps.

Reply:
I do not see how `find_path!` gives you the information of which target is the "best" though...?

Reply:
It doesn't, but it was better than nothing and didn't make you dig into deeply nested undocumented properties of the pathfinder. As @AayushSabharwal suggests, we can probably export a separate method that chooses the best from multiple objectives without the need to expose `find_path`.

Reply:
ah okay I do see it. Okay `find_path` is part of the API your example made it clear to me

Reply:
we should mention exactly this exacmple as an example block in the docstring

Reply:
I'll add the function now, but I don't think I'll mark this as resolved till it's properly documented